presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -583,6 +663,139 @@ public Void visitCall(CallExpression call, Context context)
+                        RowType rowType = (RowType) ((ArrayType) expression.getType()).getElementType();
+                        List<Field> sortedFields = sortedCopyOf((x, y) -> {
+                            if (x.getType() instanceof FixedWidthType && !(y.getType() instanceof FixedWidthType)) {
+                                return -1;
+                            }
+                            else if (y.getType() instanceof FixedWidthType && !(x.getType() instanceof FixedWidthType)) {
+                                return 1;
+                            }
+                            else if (!(x.getType() instanceof FixedWidthType) && !(y.getType() instanceof FixedWidthType)) {
+                                return 0;
+                            }
+                            return Integer.compare(((FixedWidthType) x.getType()).getFixedSize(), ((FixedWidthType) y.getType()).getFixedSize());
+                        }, rowType.getFields());
+
+                        Optional<Field> smallestField = sortedFields.stream().filter(x -> x.getName().isPresent()).findFirst();
+                        if (smallestField.isPresent()) {
+                            List<PathElement> newPath = ImmutableList.<PathElement>builder()
+                                    .addAll(subfield.get().getPath())
+                                    .addAll(prefixPath)
+                                    .add(nestedField(smallestField.get().getName().get()))
+                                    .build();
+                            context.subfields.add(new Subfield(subfield.get().getRootName(), newPath));
+                        }
+                        return true;
+                    }
+                    if (subfield.get().getPath().isEmpty() && expression instanceof VariableReferenceExpression) {
+                        context.variables.add((VariableReferenceExpression) expression);
+                    }
+                    else {
+                        context.subfields.add(subfield.get());
+                    }
+                }
+
+                for (Subfield accessedArrayElementSubfield : accessedSubfields) {
+                    List<PathElement> newPath = ImmutableList.<PathElement>builder()
+                            .addAll(subfield.get().getPath())
+                            .addAll(prefixPath)
+                            .addAll(accessedArrayElementSubfield.getPath())
+                            .build();
+                    context.subfields.add(new Subfield(subfield.get().getRootName(), newPath));
+                }
+                return true;
+            }
+
+            private boolean isPushDownSubfieldsFromLambdasEligible(CallExpression call)
+            {
+                QualifiedObjectName functionName = getFunctionName(call.getFunctionHandle());
+                if (call.getArguments().get(0).getType() instanceof ArrayType &&
```

Comment:
I wonder if this can be split for readability:
```
if (!call.getArguments().get(0).getType() instanceof ArrayType ||
        !arrayFunctionEligibleForSubfieldPruning.containsKey(functionName)) {
        return false;
}
return !arrayFunctionEligibleForSubfieldPruning.get(functionName).isOutputtingAllSubfields();
```